Broken PoolingHttpClientConnectionManager causing "Received fatal alert: bad_certificate" on Bot Certificate Authentication

## Bug Report
🐛 

### Steps to Reproduce:
1. Have imported the right trust store
2. Using bot certificate authentication
3. Run bot as usual

### Expected Result:
Bot should run without any issue

### Actual Result:

Unable to run the bot with the certificate authentication.
Stacktrace :
`Caused by: java.lang.RuntimeException: Network error occurred while trying to connect to the "SESSION_AUTH" at the following address: https://---/sessionauth. Error while trying to validate certificate for the trust store. This type of error typically means that your network is using a self-signed certificate.
	at com.symphony.bdk.core.auth.impl.AuthenticationRetry.executeAndRetry(AuthenticationRetry.java:92)
	at com.symphony.bdk.core.auth.impl.AbstractBotAuthenticator.retrieveToken(AbstractBotAuthenticator.java:28)
	at com.symphony.bdk.core.auth.impl.BotAuthenticatorCertImpl.retrieveSessionToken(BotAuthenticatorCertImpl.java:54)
	at com.symphony.bdk.core.auth.impl.AuthSessionCertImpl.refresh(AuthSessionCertImpl.java:44)
	at com.symphony.bdk.core.auth.impl.BotAuthenticatorCertImpl.authenticateBot(BotAuthenticatorCertImpl.java:47)
	at com.symphony.bdk.core.SymphonyBdk.<init>(SymphonyBdk.java:110)
	at com.symphony.bdk.core.SymphonyBdk.<init>(SymphonyBdk.java:85)
	at flo.iflow.bot.engine.service.SymphonyBotService.init(SymphonyBotService.java:59)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:389)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:333)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:157)
	... 31 common frames omitted
Caused by: javax.ws.rs.ProcessingException: javax.net.ssl.SSLHandshakeException: Received fatal alert: bad_certificate`

Log from javax.net.debug : 
`
javax.net.ssl|FINE|01|main|2021-05-25 13:36:13.569 SGT|CertificateMessage.java:297|No X.509 certificate for client authentication, use empty Certificate message instead
javax.net.ssl|FINE|01|main|2021-05-25 13:36:13.569 SGT|CertificateMessage.java:328|Produced client Certificate handshake message (
"Certificates": <empty list>
)`

### Environment:
`>= BDK 2.1.3`

### Additional Context:
The stacktrace above is quite misleading also, saying there is an issue with our trust store, but in fact nothing wrong with our trust store.
The recent changes by introducing `PoolingHttpClientConnectionManager` on `ApiClientBuilderJersey2 ` is breaking SSL bot certificate authentication. 
The pool is constructed without proper SSLContext, it causes the private certificate file which configured by the symphony bdk is not being pickup during the API invocation.
Reference :
https://github.com/finos/symphony-bdk-java/pull/430

Hi @wngasinur, 

Certificate authentication is still working fine on our side with latest [2.1.8](https://github.com/finos/symphony-bdk-java/releases/tag/2.1.8) release. 

We however managed to reproduce your issue when revoking the CA from ACP: 
![Screen Shot 2021-06-11 at 16 51 38](https://user-images.githubusercontent.com/39826516/121705875-623de380-cad5-11eb-8041-51fde7043966.png)
Here's the stacktrace I obtained: 
```
16:49:38.609 [main] ERROR com.symphony.bdk.core.retry.RetryWithRecovery - Network error occurred while trying to connect to the "SESSION_AUTH" at the following address: https://localhost.symphony.com:8444/sessionauth. Error while trying to validate certificate for the trust store. This type of error typically means that your network is using a self-signed certificate.javax.net.ssl.SSLHandshakeException: Received fatal alert: certificate_unknown
Exception in thread "main" java.lang.RuntimeException: Network error occurred while trying to connect to the "SESSION_AUTH" at the following address: https://localhost.symphony.com:8444/sessionauth. Error while trying to validate certificate for the trust store. This type of error typically means that your network is using a self-signed certificate.
	at com.symphony.bdk.core.auth.impl.AuthenticationRetry.executeAndRetry(AuthenticationRetry.java:92)
	at com.symphony.bdk.core.auth.impl.AbstractBotAuthenticator.retrieveToken(AbstractBotAuthenticator.java:28)
	at com.symphony.bdk.core.auth.impl.BotAuthenticatorCertImpl.retrieveSessionToken(BotAuthenticatorCertImpl.java:54)
	at com.symphony.bdk.core.auth.impl.AuthSessionCertImpl.refresh(AuthSessionCertImpl.java:44)
	at com.symphony.bdk.core.auth.impl.BotAuthenticatorCertImpl.authenticateBot(BotAuthenticatorCertImpl.java:47)
	at com.symphony.bdk.core.SymphonyBdk.<init>(SymphonyBdk.java:110)
	at com.symphony.bdk.core.SymphonyBdk.<init>(SymphonyBdk.java:85)
	at com.symphony.bdk.BotApplication.main(BotApplication.java:27)
Caused by: javax.ws.rs.ProcessingException: javax.net.ssl.SSLHandshakeException: Received fatal alert: certificate_unknown
	at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:268)
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:297)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$0(JerseyInvocation.java:662)
	at org.glassfish.jersey.client.JerseyInvocation.call(JerseyInvocation.java:697)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$runInScope$3(JerseyInvocation.java:691)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
	at org.glassfish.jersey.client.JerseyInvocation.runInScope(JerseyInvocation.java:691)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:661)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:439)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:345)
	at com.symphony.bdk.http.jersey2.ApiClientJersey2.getResponse(ApiClientJersey2.java:179)
	at com.symphony.bdk.http.jersey2.ApiClientJersey2.invokeAPI(ApiClientJersey2.java:135)
	at com.symphony.bdk.gen.api.CertificateAuthenticationApi.v1AuthenticatePostWithHttpInfo(CertificateAuthenticationApi.java:411)
	at com.symphony.bdk.gen.api.CertificateAuthenticationApi.v1AuthenticatePost(CertificateAuthenticationApi.java:364)
	at com.symphony.bdk.core.auth.impl.BotAuthenticatorCertImpl.authenticateAndGetToken(BotAuthenticatorCertImpl.java:65)
	at com.symphony.bdk.core.auth.impl.AbstractBotAuthenticator.lambda$retrieveToken$0(AbstractBotAuthenticator.java:29)
	at com.symphony.bdk.core.retry.RetryWithRecovery.executeOnce(RetryWithRecovery.java:105)
	at io.github.resilience4j.retry.Retry.lambda$decorateCheckedSupplier$3f69f149$1(Retry.java:137)
	at io.github.resilience4j.retry.Retry.executeCheckedSupplier(Retry.java:419)
	at com.symphony.bdk.core.retry.resilience4j.Resilience4jRetryWithRecovery.execute(Resilience4jRetryWithRecovery.java:65)
	at com.symphony.bdk.core.auth.impl.AuthenticationRetry.executeAndRetry(AuthenticationRetry.java:85)
	... 7 more
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: certificate_unknown
	at sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at sun.security.ssl.Alert.createSSLException(Alert.java:117)
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:311)
	at sun.security.ssl.Alert$AlertConsumer.consume(Alert.java:293)
	at sun.security.ssl.TransportContext.dispatch(TransportContext.java:185)
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:149)
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1143)
	at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1054)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:394)
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1570)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:352)
	at org.glassfish.jersey.client.internal.HttpUrlConnector._apply(HttpUrlConnector.java:377)
	at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:266)
	... 30 more
```
Could you please verify if the certificate you are using is correct?

--- a/symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java
+++ b/symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2.java
@@ -6,6 +6,11 @@ import com.symphony.bdk.http.api.ApiClient;
 import com.symphony.bdk.http.api.ApiClientBuilder;
 import com.symphony.bdk.http.api.util.ApiUtils;
 
+import org.apache.http.config.Registry;
+import org.apache.http.config.RegistryBuilder;
+import org.apache.http.conn.socket.ConnectionSocketFactory;
+import org.apache.http.conn.socket.PlainConnectionSocketFactory;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
 import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
 import org.apiguardian.api.API;
 import org.glassfish.jersey.SslConfigurator;
@@ -82,9 +87,10 @@ public class ApiClientBuilderJersey2 implements ApiClientBuilder {
   public ApiClient build() {
     java.util.logging.Logger.getLogger("org.glassfish.jersey.client").setLevel(java.util.logging.Level.SEVERE);
 
+    SSLContext sslContext = this.createSSLContext();
     final Client httpClient = ClientBuilder.newBuilder()
-        .sslContext(this.createSSLContext())
-        .withConfig(this.createClientConfig())
+        .sslContext(sslContext)
+        .withConfig(this.createClientConfig(sslContext))
         .build();
 
     httpClient.property(ClientProperties.CONNECT_TIMEOUT, this.connectionTimeout);
@@ -205,7 +211,7 @@ public class ApiClientBuilderJersey2 implements ApiClientBuilder {
   }
 
   @API(status = API.Status.EXPERIMENTAL)
-  protected ClientConfig createClientConfig() {
+  protected ClientConfig createClientConfig(SSLContext sslContext) {
     final ClientConfig clientConfig = new ClientConfig();
     this.configureJackson(clientConfig);
     if (this.proxyUrl != null) {
@@ -217,12 +223,20 @@ public class ApiClientBuilderJersey2 implements ApiClientBuilder {
     clientConfig.property(HttpUrlConnectorProvider.SET_METHOD_WORKAROUND, true);
     // turn off compliance validation to be able to send payloads with DELETE calls
     clientConfig.property(ClientProperties.SUPPRESS_HTTP_COMPLIANCE_VALIDATION, true);
+
+    SSLConnectionSocketFactory sslConnectionSocketFactory = new SSLConnectionSocketFactory(sslContext);
+    Registry<ConnectionSocketFactory> registry = RegistryBuilder.<ConnectionSocketFactory>create()
+        .register("https", sslConnectionSocketFactory)
+        .register("http", new PlainConnectionSocketFactory())
+        .build();
+
     // By default PoolingHttpClientConnectionManager, if not configured, has 20 connection in the
     // pool BUT only 2 max connection per route.
-    final PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+    final PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager(registry);
     connectionManager.setMaxTotal(this.connectionPoolMax);
     connectionManager.setDefaultMaxPerRoute(this.connectionPoolPerRoute);
     clientConfig.property(ApacheClientProperties.CONNECTION_MANAGER, connectionManager);
+    clientConfig.connectorProvider(new ApacheConnectorProvider());
     return clientConfig;
   }
 
@@ -234,7 +248,6 @@ public class ApiClientBuilderJersey2 implements ApiClientBuilder {
 
   @API(status = API.Status.EXPERIMENTAL)
   protected void configureProxy(ClientConfig clientConfig) {
-    clientConfig.connectorProvider(new ApacheConnectorProvider());
     clientConfig.property(ClientProperties.PROXY_URI, proxyUrl);
     clientConfig.property(ClientProperties.PROXY_USERNAME, proxyUser);
     clientConfig.property(ClientProperties.PROXY_PASSWORD, proxyPassword);

--- a/symphony-bdk-http/symphony-bdk-http-jersey2/src/test/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2Test.java
+++ b/symphony-bdk-http/symphony-bdk-http-jersey2/src/test/java/com/symphony/bdk/http/jersey2/ApiClientBuilderJersey2Test.java
@@ -1,0 +1,73 @@
+package com.symphony.bdk.http.jersey2;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.symphony.bdk.http.api.ApiClient;
+import com.symphony.bdk.http.api.ApiException;
+import com.symphony.bdk.http.api.ApiResponse;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockserver.configuration.ConfigurationProperties;
+import org.mockserver.integration.ClientAndServer;
+import org.mockserver.logging.MockServerLogger;
+import org.mockserver.model.HttpRequest;
+import org.mockserver.model.HttpResponse;
+import org.mockserver.socket.tls.KeyStoreFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.util.Collections;
+
+class ApiClientBuilderJersey2Test {
+
+  private static final String KEY_STORE_PWD = "changeit";
+
+  private ClientAndServer mockServer;
+
+  @BeforeEach
+  void setUp() {
+    // enforce client certificates usage
+    ConfigurationProperties.tlsMutualAuthenticationRequired(true);
+    mockServer = ClientAndServer.startClientAndServer();
+  }
+
+  @AfterEach
+  void tearDown() {
+    mockServer.stop();
+  }
+
+  @Test
+  void sslContextIsUsed()
+      throws ApiException, CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
+    ByteArrayOutputStream keyStoreData = getMockServerKeyStore();
+    ApiClient client = new ApiClientBuilderJersey2()
+        .withBasePath("https://localhost:" + mockServer.getPort())
+        .withKeyStore(keyStoreData.toByteArray(), "changeit")
+        .withTrustStore(keyStoreData.toByteArray(), "changeit")
+        .build();
+
+    mockServer.withSecure(true)
+        .when(HttpRequest.request().withMethod("GET").withPath("/test"))
+        .respond(HttpResponse.response().withStatusCode(200));
+
+    ApiResponse<Object> response =
+        client.invokeAPI("/test", "GET", Collections.emptyList(), null, Collections.emptyMap(), Collections.emptyMap(),
+            null, "application/json", "", null, null);
+
+    assertEquals(200, response.getStatusCode());
+  }
+
+  private ByteArrayOutputStream getMockServerKeyStore()
+      throws KeyStoreException, IOException, NoSuchAlgorithmException, CertificateException {
+    KeyStore mockServerKeyStore = new KeyStoreFactory(new MockServerLogger()).loadOrCreateKeyStore();
+    ByteArrayOutputStream keyStoreData = new ByteArrayOutputStream();
+    mockServerKeyStore.store(keyStoreData, KEY_STORE_PWD.toCharArray());
+    return keyStoreData;
+  }
+}

